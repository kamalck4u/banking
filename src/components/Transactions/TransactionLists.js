import React from 'react';
import { useHistory } from 'react-router-dom';
import { REFERENCE_NUMBER, DATE, AMOUNT, LOADER_TEXT_TRANSACTIONS } from '../Common/Constants';
import { Loader } from '../Common/Panels';
import '../../App.css';
import { array, bool } from 'prop-types';

const TransactionLists = ({ transactions, isLoading }) => {
	const history = useHistory();
	return (
		<React.Fragment>
			{
				isLoading ? <Loader loaderText={LOADER_TEXT_TRANSACTIONS} /> :
					<div className="table-responsive">
						<table className="table table-striped">
							<thead className="thead-dark">
								<tr>
									<th scope="col">{REFERENCE_NUMBER}</th>
									<th scope="col">{DATE}</th>
									<th scope="col">{AMOUNT}</th>
								</tr>
							</thead>
							<tbody className="row-clickable">
								{transactions.map((transaction) => {
									return (
										<tr key={transaction.id} className="clickable-row table-cell" onClick={() => history.push(`/transactions/${transaction.id}/details`, { transaction })}>
											<td>{transaction.id}</td>
											<td>{transaction.transaction_date}</td>
											<td>{transaction.amount}</td>
										</tr>
									);
								}
								)}
							</tbody>
						</table>
					</div>
			}
		</React.Fragment>
	);
};

TransactionLists.propTypes = {
	transactions: array,
	isLoading: bool
};

export default TransactionLists;


