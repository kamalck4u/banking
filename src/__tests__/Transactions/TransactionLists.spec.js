import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import TransactionLists from '../../components/Transactions/TransactionLists';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const transactions = [{
	'id': 1,
	'account_id': 1,
	'description': 'Dolore quis ad et mollit nisi excepteur ex anim fugiat quis ipsum exercitation proident cupidatat. Quis anim incididunt excepteur cupidatat aliquip nulla reprehenderit velit. Dolor pariatur velit consectetur dolore aliquip reprehenderit non aliqua consectetur. Sunt aliquip consequat et in eu aute.\r\n',
	'from': 'Savings Account - 342423455344',
	'transaction_date': '2019-06-08T03:37:28 -08:00',
	'transaction_processed': true,
	'amount': '$1,000.10'
}];

let isloading = false;
const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useHistory: () => ({
		push: mockHistoryPush,
	}),
}));

describe('TransactionLists', () => {
	it('snapshot renders', () => {
		const TransactionListsComponent = renderer.create(<TransactionLists transactions={transactions} isLoading={isloading} />);
		expect(TransactionListsComponent.toJSON()).toMatchSnapshot();
	});
});

describe('Transaction List', () => {
	let wrapper;
	beforeEach(() => {
		wrapper = shallow(<TransactionLists transactions={transactions} isLoading={isloading} />);
	});

	it('should show the table headers', () => {
		expect(wrapper.find('th')).toHaveLength(3);
		expect(wrapper.find('th').getElements()[0].props.children).toEqual('Transaction#');
		expect(wrapper.find('th').getElements()[1].props.children).toEqual('Date');
		expect(wrapper.find('th').getElements()[2].props.children).toEqual('Amount');

	});

	it('should render the table rows', () => {
		expect(wrapper.find('td')).toHaveLength(3);
		expect(wrapper.find('td').getElements()[0].props.children).toEqual(1);
		expect(wrapper.find('td').getElements()[1].props.children).toEqual('2019-06-08T03:37:28 -08:00');
		expect(wrapper.find('td').getElements()[2].props.children).toEqual('$1,000.10');
	});

	it('should be clickable', () => {
		expect(wrapper.find('tbody.row-clickable')).toHaveLength(1);
		expect(wrapper.find('tr.clickable-row.table-cell')).toHaveLength(1);
	});

	it('should redirect to transaction Details page', () => {
		wrapper.find('tr').at(1).simulate('click');
		expect(mockHistoryPush).toHaveBeenCalledWith('/transactions/1/details', {
			transaction: {
				'id': 1,
				'account_id': 1,
				'description': 'Dolore quis ad et mollit nisi excepteur ex anim fugiat quis ipsum exercitation proident cupidatat. Quis anim incididunt excepteur cupidatat aliquip nulla reprehenderit velit. Dolor pariatur velit consectetur dolore aliquip reprehenderit non aliqua consectetur. Sunt aliquip consequat et in eu aute.\r\n',
				'from': 'Savings Account - 342423455344',
				'transaction_date': '2019-06-08T03:37:28 -08:00',
				'transaction_processed': true,
				'amount': '$1,000.10'
			}
		});
	});
});

describe('Panel', () => {
	let wrapper;
	it('should show spinner while data is being fetched', () => {
		isloading = true;
		wrapper = shallow(<TransactionLists transactions={transactions} isLoading={isloading} />);
		expect(wrapper.find('Loader').props()).toEqual({ loaderText: 'Your transactions are being loaded, please wait...' });
		expect(wrapper.find('th')).toHaveLength(0);
	});
});

