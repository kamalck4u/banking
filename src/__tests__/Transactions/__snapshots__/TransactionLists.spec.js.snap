// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionLists snapshot renders 1`] = `
<div
  className="table-responsive"
>
  <table
    className="table table-striped"
  >
    <thead
      className="thead-dark"
    >
      <tr>
        <th
          scope="col"
        >
          Transaction#
        </th>
        <th
          scope="col"
        >
          Date
        </th>
        <th
          scope="col"
        >
          Amount
        </th>
      </tr>
    </thead>
    <tbody
      className="row-clickable"
    >
      <tr
        className="clickable-row table-cell"
        onClick={[Function]}
      >
        <td>
          1
        </td>
        <td>
          2019-06-08T03:37:28 -08:00
        </td>
        <td>
          $1,000.10
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
